arbnode/message-extraction/extraction-function/batch_serialization.go

```diff
@@ -0,0 +1,125 @@
+package extractionfunction
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth/filters"
+
+	"github.com/offchainlabs/nitro/arbnode"
+	"github.com/offchainlabs/nitro/daprovider"
+	"github.com/offchainlabs/nitro/solgen/go/bridgegen"
+)
+
+func serializeBatch(
+	ctx context.Context,
+	batch *arbnode.SequencerInboxBatch,
+	tx *types.Transaction,
+	txIndex uint,
+	receiptFetcher ReceiptFetcher,
+) ([]byte, error) {
+	if batch.Serialized != nil {
+		return batch.Serialized, nil
+	}
+
+	var fullData []byte
+
+	// Serialize the header
+	headerVals := []uint64{
+		batch.TimeBounds.MinTimestamp,
+		batch.TimeBounds.MaxTimestamp,
+		batch.TimeBounds.MinBlockNumber,
+		batch.TimeBounds.MaxBlockNumber,
+		batch.AfterDelayedCount,
+	}
+	for _, bound := range headerVals {
+		var intData [8]byte
+		binary.BigEndian.PutUint64(intData[:], bound)
+		fullData = append(fullData, intData[:]...)
+	}
+
+	// Append the batch data
+	data, err := getSequencerBatchData(
+		ctx,
+		batch,
+		tx,
+		txIndex,
+		receiptFetcher,
+	)
+	if err != nil {
+		return nil, err
+	}
+	fullData = append(fullData, data...)
+
+	batch.Serialized = fullData
+	return fullData, nil
+}
+
+func getSequencerBatchData(
+	ctx context.Context,
+	batch *arbnode.SequencerInboxBatch,
+	tx *types.Transaction,
+	txIndex uint,
+	receiptFetcher ReceiptFetcher,
+) ([]byte, error) {
+	addSequencerL2BatchFromOriginCallABI := seqInboxABI.Methods["addSequencerL2BatchFromOrigin0"]
+	switch batch.DataLocation {
+	case arbnode.BatchDataTxInput:
+		data := tx.Data()
+		if len(data) < 4 {
```

Comment:
curiosity: not sure why 4 is the threshold used here

Reply:
the first 4 bytes of a transaction data are the method selector hash, such as keccak256(payBalance(uint256, address))[:4]